src/ucp/dt/dt_struct.c

```diff
@@ -0,0 +1,365 @@
+{
+    ssize_t eidx = -1;
+    size_t elem_len = 0, copy_len = 0;
+    size_t in_offset = in_offset_orig;
+    size_t in_offset_rel = 0, rep_num = 0;
+    ptrdiff_t out_offset = 0;
+    ucp_dt_struct_t *sub_s;
+
+    /* Seek for the offset */
+    eidx = _elem_by_offset(s, in_offset, &in_offset_rel, &rep_num);
+
+    while( (0 < len) && rep_num < s->rep_count){
+        ucp_struct_dt_desc_t *dsc = &s->desc[eidx];
+        out_offset = dsc->displ + dsc->extent * rep_num;
+        switch (dsc->dt & UCP_DATATYPE_CLASS_MASK) {
+        case UCP_DATATYPE_CONTIG:
+            elem_len = ucp_contig_dt_length(dsc->dt, 1);
+            copy_len = ucs_min(elem_len - in_offset_rel, len);
+            memcpy(outbuf + out_offset, inbuf + in_offset, copy_len);
+            break;
+        case UCP_DATATYPE_STRUCT:
+            sub_s = ucp_dt_struct(dsc->dt);
+            copy_len = _dte_unpack(sub_s, inbuf + in_offset, outbuf + out_offset,
+                      in_offset_rel, len);
+            break;
+        }
+        /* after the first iteration we will always be copying from the
+         * beginning of each structural element
+         */
+        in_offset += copy_len;
+        len -= copy_len;
+        in_offset_rel = 0;
+        eidx++;
+        if(!(eidx < s->desc_count)) {
+            eidx = 0;
+            rep_num++;
+        }
+    }
+
+    /* Return processed length */
+    return (in_offset - in_offset_orig);
+}
+
+ucs_status_t ucp_dt_create_struct(ucp_struct_dt_desc_t *desc_ptr,
+                                  size_t desc_count, size_t rep_count,
+                                  ucp_datatype_t *datatype_p)
+{
+    ucp_dt_struct_t *dt;
+    int ret;
+    size_t i;
+
+    for(i=0; i < desc_count; i++){
+        switch (desc_ptr[i].dt & UCP_DATATYPE_CLASS_MASK) {
+        case UCP_DATATYPE_CONTIG:
+        case UCP_DATATYPE_STRUCT:
+            /* OK */
+            break;
+        case UCP_DATATYPE_IOV:
+        case UCP_DATATYPE_GENERIC:
+            /* Not supported */
+            return UCS_ERR_NOT_IMPLEMENTED;
+        }
+    }
+
+    /* Sanity check:
+     * Structured datatype only supports UCP_DATATYPE_CONTIG and
+     * UCP_DATATYPE_STRUCT as sub-datatypes
+     */
+
+    ret = ucs_posix_memalign((void **)&dt,
+                             ucs_max(sizeof(void *), UCS_BIT(UCP_DATATYPE_SHIFT)),
+                             sizeof(*dt), "struct_dt");
+    if (ret != 0) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    ret = ucs_posix_memalign((void **)&dt->desc, sizeof(*dt->desc),
+                             sizeof(*dt->desc) * dt->desc_count,
+                             "ucp_dt_struct_t");
+    if (ret != 0) {
+        ucs_free(dt);
+        return UCS_ERR_NO_MEMORY;
+    }
+    memcpy(dt->desc, desc_ptr, sizeof(*desc_ptr) * desc_count);
+    dt->desc_count = desc_count;
+    dt->rep_count = rep_count;
+    _set_length(dt);
+    _set_depth(dt);
+    *datatype_p = ((uintptr_t)dt) | UCP_DATATYPE_STRUCT;
```

Comment:
would need to init hash here (if more than 1 element, or single but non-contig element)